udata/api_fields.py

```diff
@@ -0,0 +1,239 @@
+            parser.add_argument('page_size', type=int, location='args', default=20, help='The page size')
+        
+        if sortables:
+            choices = sortables + ['-' + k for k in sortables]
+            parser.add_argument('sort', type=str, location='args', choices=choices, help='The field (and direction) on which sorting apply')
+
+        cls.__index_parser__ = parser
+        def apply_sort_filters_and_pagination(base_query):
+            args = cls.__index_parser__.parse_args()
+
+            if sortables and args['sort']:
+                base_query = base_query.order_by(args['sort'])
+
+            if paginable:
+                base_query = base_query.paginate(args['page'], args['page_size'])
+
+            return base_query
+
+        cls.apply_sort_filters_and_pagination = apply_sort_filters_and_pagination
+        return cls
+    return wrapper
+
+
+def function_field(**info):
+    def inner(func):
+        func.__additional_field_info__ = info
+        return func
+
+    return inner
+
+def field(inner, **kwargs):
+    '''
+    Simple decorator to mark a field as visible for the API fields.
+    We can pass additional arguments that will be forward to the RestX field constructor.
+    '''
+    inner.__additional_field_info__ = kwargs
+    return inner
+
+
+def patch(obj, request): 
+    '''
+    Patch the object with the data from the request.
+    Only fields decorated with the `field()` decorator will be read (and not readonly).
+    '''
+    for key, value in request.json.items():
+        field = obj.__write_fields__.get(key)
+        if field is not None and not field.readonly:
+            model_attribute = getattr(obj.__class__, key)
+            if isinstance(model_attribute, mongoengine.fields.ListField) and isinstance(model_attribute.field, mongoengine.fields.ReferenceField):
+                # TODO `wrap_primary_key` do Mongo request, do a first pass to fetch all documents before calling it (to avoid multiple queries).
+                value = [wrap_primary_key(key, model_attribute.field, id) for id in value]
+            if isinstance(model_attribute, mongoengine.fields.ReferenceField):
+                value = wrap_primary_key(key, model_attribute, value)
+
+            setattr(obj, key, value)
+
+    return obj
+
+def wrap_primary_key(field_name: str, foreign_field: mongoengine.fields.ReferenceField, value: str):
+    '''
+    We need to wrap the `String` inside an `ObjectId` most of the time. If the foreign ID is a `String` we need to get 
+    a `DBRef` from the database.
+
+    TODO: we only check the document reference if the ID is a `String` field (not in the case of a classic `ObjectId`).
+    '''
+    document_type = foreign_field.document_type()
+    id_field_name = document_type.__class__._meta["id_field"]
+
+    id_field = getattr(document_type.__class__, id_field_name)
+
+    if isinstance(id_field, mongoengine.fields.ObjectIdField):
+        return ObjectId(value)
```

Comment:
How to make sure that the object with this id exists?

Reply:
We could do the same thing as below, I first did it below to get a DBRef more than to test the existance (because I think MongoDB will check this, but not sure, I’ll add a test.)

Reply:
Done in fd99086fc (MongoDB didn't check before-hand the validity of the reference, it only crash on read)